File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.go
func min(a, b int) int {
  if a > b {
    return b
  } else {
    return a
  }
}
func max(a, b int) int {
  if a > b {
    return a
  } else {
    return b
  }
}
func maxProfit(prices []int) int {
  dp := math.MaxInt32
  ans := 0
  for _, v := range prices {
    dp = min(dp, v)
    ans = max(ans, v - dp)
  }
  return ans
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.rs
impl Solution {
  pub fn max_profit(prices: Vec<i32>) -> i32 {
    let mut dp = prices[0];
    let mut ans = 0;
    for p in prices {
      dp = dp.min(p);
      ans = ans.max(p - dp);
    }
    return ans;
  }
}